wgengine/magicsock/magicsock.go

```diff
@@ -1239,7 +1246,38 @@ func (c *Conn) GetEndpointChanges(peer tailcfg.NodeView) ([]EndpointChange, erro
 
 // DiscoPublicKey returns the discovery public key.
 func (c *Conn) DiscoPublicKey() key.DiscoPublic {
-	return c.discoPublic
+	return c.discoPublic.Load()
+}
+
+// RotateDiscoKey generates a new discovery key pair and updates the connection
+// to use it. This invalidates all existing disco sessions and will cause peers
+// to re-establish discovery sessions with the new key.
+//
+// This method is primarily intended for testing and debugging.
+//
+// Rotations are serialized to ensure only one rotation happens at a time.
+func (c *Conn) RotateDiscoKey() {
+	c.rotateDiscoMu.Lock()
+	defer c.rotateDiscoMu.Unlock()
+
+	oldShort := c.discoShort.Load()
+	newPrivate := key.NewDisco()
+	newPublic := newPrivate.Public()
+	newShort := newPublic.ShortString()
+
+	c.discoPrivate.Store(newPrivate)
+	c.discoPublic.Store(newPublic)
```

Comment:
Concern (blocking): Since we never acquire the rotate mutex when reading, isn't there a chance that these won't match? Does that matter? I don't understand the implications of mismatched private and public keys.

Reply:
If reader code didn't hold the lock over a long range it could easily be reading disjoint keys between reads. Sadly we really need a larger state machine overhaul to be completely sure of this.

I can add a defensive type that ensures updates to these are atomic and provides an API to recover both in a single atomic read, but this still doesn't produce safety if the calling code is unsound.

fwiw there are lots of toctou style errors in the code in this regard, for example most of the log statements don't log the key that was used to present a message, they log the key that is currently in state.

Reply:
I introduced a new type that would provide a way to get both keys in a synchronized fashion. There's essentially no usage of it as they're never really used together, but it's now present.

Reply:
We should at least give the calling code a fighting chance to get things right. Thanks for addressing this.